moveit_ros/moveit_servo/src/command_processor.cpp

```diff
@@ -0,0 +1,270 @@
+}
+
+namespace moveit_servo
+{
+
+CommandProcessor::CommandProcessor(const planning_scene_monitor::PlanningSceneMonitorPtr& planning_scene_monitor,
+                                   const moveit::core::JointModelGroup* joint_model_group,
+                                   moveit::core::RobotStatePtr& current_state, servo::Params& servo_params,
+                                   StatusCode& servo_status)
+  : planning_scene_monitor_{ planning_scene_monitor }
+  , joint_model_group_{ joint_model_group }
+  , robot_state_{ current_state }
+  , servo_params_{ servo_params }
+  , servo_status_{ servo_status }
+{
+  num_joints_ = joint_model_group_->getActiveJointModelNames().size();
+
+  // Create PID controllers for pose tracking
+  rclcpp::WallRate controller_rate(1.0 / servo_params_.publish_period);
+  controller_period_ = controller_rate.period().count();
+  controllers_ = createControllers(servo_params_);
+  RCLCPP_INFO_STREAM(LOGGER, "PID controllers created.");
+
+  setIKSolver();
+}
+
+Eigen::VectorXd CommandProcessor::jointDeltaFromCommand(const JointJog& command)
+{
+  // Find the target joint position based on the commanded joint velocity
+  Eigen::VectorXd joint_position_delta(num_joints_);
+  joint_position_delta.setZero();
+
+  if (isValidCommand(command))
+  {
+    // The incoming command should be in rad/s
+    joint_position_delta = command * servo_params_.publish_period;
+  }
+  else
+  {
+    servo_status_ = StatusCode::INVALID;
+    RCLCPP_WARN_STREAM(LOGGER, "Invalid joint velocity command");
+  }
+  return joint_position_delta;
+}
+
+Eigen::VectorXd CommandProcessor::jointDeltaFromCommand(const Twist& command)
+{
+  Eigen::VectorXd joint_position_delta(num_joints_);
+  joint_position_delta.setZero();
+  Eigen::VectorXd cartesian_position_delta;
+
+  const bool has_transform = transformExists(robot_state_, command.frame_id);
+  const bool valid_command = isValidCommand(command.velocities);
+
+  if (has_transform && valid_command)
+  {
+    // Compute the cartesian position delta based on incoming command, assumed to be in m/s
+    cartesian_position_delta = command.velocities * servo_params_.publish_period;
+
+    // Compute the required change in joint angles.
+    if (ik_solver_)
+    {
+      // Use robot's IK solver to get joint position delta.
+      joint_position_delta = deltaFromIkSolver(cartesian_position_delta);
+    }
+    else
+    {
+      // Robot does not have an IK solver, use inverse Jacobian to compute IK.
+      Eigen::MatrixXd jacobian = robot_state_->getJacobian(joint_model_group_);
+      Eigen::JacobiSVD<Eigen::MatrixXd> svd =
+          Eigen::JacobiSVD<Eigen::MatrixXd>(jacobian, Eigen::ComputeThinU | Eigen::ComputeThinV);
+      Eigen::MatrixXd matrix_s = svd.singularValues().asDiagonal();
+      Eigen::MatrixXd pseudo_inverse = svd.matrixV() * matrix_s.inverse() * svd.matrixU().transpose();
+
+      joint_position_delta = pseudo_inverse * cartesian_position_delta;
```

Comment:
Could put this in the utilities file so this looks even cleaner

Reply:
I combined this together with the IK solver and made it into a single method called `jointDeltaFromIK`.
Looks much cleaner overall.